### PR TITLE
 Dynamically register formula when first used

### LINKED ISSUE #197: Dynamically register a formula when needed instead of front-loading all of them
On my machine, activating the BHoM takes about 45 seconds. Half the time is used to collect the information necessary to generate each formula and compile it. The other half register them through Excel-DNA. If I deactivate the front-loading of the formulas and only leave the initialisation of the global menu, the activation time drops to around 2 seconds.

Given that only a small fraction of the formula set is ever going to be used in a spreadsheet (I have 5000 with the toolkit installed on my machine), wouldn't it be better to register a formula only when they are needed? This means either when they are selected by the user or when opening a spreadsheet containing BHoM formulas.

If we do this, both method registration and intellisense will have to be activated per formula. I am not 100% sure that Excel-DNA is capable of doing that properly in its current state so that will require some investigation. But Excel-DNA is open source so we can always augment it with what we need or try with NetOffice Excel API (that we are already using anyway).

There is also the problem of loosing the auto-completion in the formula bar. To be honest, I don't like the way it works at the moment and find it quite hard to find what I need. Currently, we are also loosing the weighting of the search items (keeping most relevant items at the top) so it makes it even less familiar than what we are used to when searching the BHoM. So I would rather have Ctrl-Shift-B working in the formula bar as well and completely ignore the auto-completion provided by Excel.

@al-fisher  @rwemay  @IsakNaslundBh   FYI

### LINKED ISSUE #202: Add support for Versioning
#### Description:

Solution is likely related to #197 in that it will hinge on changing the way that the formulas are loaded in the first place and additionally modifying the formulas in cells that use methods that need upgrading.

Some thoughts on this:

- With #197 we'll have `ComponentRequest`s serialised, with the aim of only loading, in the first instance, the formulas corresponding to them.
- With the `ComponentRequest` we should be able detect cases when the "component" it would generate no longer exists and needs to run through the versioning toolkit to upgrade.
- At present the "components" themselves are in no way serialised so we can't leverage that, rather we may need to do something like register a placeholder formula for the old method (with the same name) that simply replaces itself with the new formula, rearranging its inputs accordingly.

@adecler, @IsakNaslundBh thoughts

### PR DESCRIPTION
**Pull request by awakeman**
### Issues addressed by this PR

Closes #197 
Closes #202 

Dynamically registers the formula when selected from CTRL+SHIFT+B menu or ribbon menu and serialises used Caller into hidden sheet.

Adds versioning by virtue of serialising through `Caller.Write()`/`Caller.Read()`

### Additional comments

Tasks needed to make this PR ready
- [x] Stop registering all formulas on load
- [x] Register method when its selected from the menu
- [x] Serialise each method on use
- [x] Load used methods on loading saved file based on serialised methods
- [x] Pass through versioning when deserialising
- [x] Detect upgraded components and register a dummy formula to upgrade it
- [x] Clean out entry in register for old method when upgraded
- [x] Load full BHoM when register doesn't exist but BHoM was used (upgrade old workbooks)

### CODE REVIEW

**Comment by awakeman** [comment]
@IsakNaslundBh This is now ready for review, testing on your existing sheets should one time load all methods and (if saved) on future loads only the methods that are in use.

**Comment by awakeman** [comment]
Some of this work has increased my awareness of trying to reduce the use of COM Interop over the C APIs, much of the tasks that use the COM API could be refactored, raising an issue to look into this going forward. #209

**Comment by awakeman** [comment]
Adding do not merge to this because, in my opinion, it carries too much risk to rush into including it in the upcoming beta. Will lift this on 30th March.

**Comment by IsakNaslundBh** [comment]
@awakeman , Still have some issues loading up excel files created before these changes:

See https://burohappold.sharepoint.com/:f:/s/BHoM/EtPmoptuJq1Hmvm2C44eXCQBHcznPACr62Vny6IZq7-aIg?e=VTcezX

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Issue with loading pre-existing files resolved, issue was I had opened them in a build from before this PR was ready. Now seem to be working as expected, being able to open old sheets with things loading correctly, and can use them again after.

Some other issues found:

1. Create enum menu is empty:

image


2. Tweaks to enum (adding an item) leads to com call crash for the validation. This can be tested by storing a pushtype from before and after the following PR https://github.com/BHoM/BHoM_Adapter/pull/205

**Comment by awakeman** [comment]
Issue #187 has been made worse (occurs most time) so needs resolving as part of this PR.

**Comment by rwemay** [comment]
Great to see this thanks @awakeman . I've added a file to the meta tests - link below. Similar to @IsakNaslundBh , I can open old sheets, but am having some issues. Specifically for example, the formula `Create.Structure.Loads.Loadcase?by_String_Int32_LoadNature()` isn't working, and intellisense not coming up for it. 

https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FExcel%5FToolkit%2F%5FMacro&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4

**Comment by awakeman** [comment]
> Similar to @IsakNaslundBh , I can open old sheets, but am having some issues. Specifically for example, the formula `Create.Structure.Loads.Loadcase?by_String_Int32_LoadNature()` isn't working, and intellisense not coming up for it.

Just to clarify you mean accessing it through `ctrl`+`shift`+`b` right.

**Comment by rwemay** [comment]
> > Similar to @IsakNaslundBh , I can open old sheets, but am having some issues. Specifically for example, the formula `Create.Structure.Loads.Loadcase?by_String_Int32_LoadNature()` isn't working, and intellisense not coming up for it.
> 
> Just to clarify you mean accessing it through `ctrl`+`shift`+`b` right.

The methods don't seem to show up in either intellisense or ctrl + shift + b - that is methods that were already in the spreadsheet previously. Methods that were not in the spreadsheet previously seem to load. Could it be an issue of ignoring methods already used in old spreadsheets?

**Comment by rwemay** [comment]
Latest issues with this PR: 
 - [x] Deprecated methods not being registered for old sheets - @awakeman need to be registered for upgrade to work. 
 - [x] Create enum and create object not working - @awakeman need to rebase from master
 - [x] Tooltips/intellisense not showing up - @awakeman rebase from master?
 - [ ] Crash on exit worse now, needs stepinto/investigation
 - [ ] Truncate argument descriptions to prevent ExcelDNA doing this and producing lots of warnings @awakeman

**Comment by rwemay** [comment]
Hi @awakeman , any joy with crash on exit? We're looking to close this PR so may separate out crash on exit. One thing that is still not working (although is ticked above) is intellisense. @IsakNaslundBh  and I have looking into this, and it seems to be due to deprecated methods. We've put a couple of sheets in the folder below, and I'll email you separately the formula/location worth testing but basically: 

 - For some methods in old spreadsheets, they aren't registered and intellisense doesn't work;
 - If you open up a fresh Excel sheet, and use ctrl+shift+b or the menu to create the same method _then_ open the sample sheet, then the method and intellisense works fine. So something to do with opening an old sheet versus registering the formula in Excel before you open said sheet. 

https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FExcel%5FToolkit%2FPull%20Requests%2FPullRequest%5F208%2DDynamically%20Register%20Formulae&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4

**Comment by rwemay** [comment]
@awakeman , with the last commit, intellisense now working. I think park the crash on exit for now as I don't think is a new issue and good to get this PR merged. I was able to recreate the enum bug, and I think @IsakNaslundBh already reported this - it was on the Apater/Pull enum rather than acoustic (apologies) below for ease of ref.

image

**Comment by rwemay** [comment]
@awakeman @IsakNaslundBh , this fixes the enum issue (although it did seem to exacerbate showing the xlDNA warning dialogue). I just tested the 'deprecated' sheet in the test script folder this morning, and it looks like the RobotAdapter is not working. I can create a new one from the Ribbon, and then the formula is registered, but it's not registering the formula from the sheet. The adapter method should not have changed between versions. Apologies, I didn't pick this up before the last commit - I didn't think it was an issue but can't see what might have caused it in last commit.

**Comment by rwemay** [comment]
Hi @IsakNaslundBh , @awakeman , any updates on the adapter issue just above, were either of you able to reproduce in the 'deprecated' spreadsheet in the test folder?

**Comment by IsakNaslundBh** [comment]
> Hi @IsakNaslundBh , @awakeman , any updates on the adapter issue just above, were either of you able to reproduce in the 'deprecated' spreadsheet in the test folder?

Not working for me either. If I add a new RobotAdapter in any other cell, the one in the initial cells kiscks in though, so looks like some missed registration to me.

**Comment by rwemay** [comment]
@awakeman , as mentioned, we need to make sure Engine.GetProperty() and Engine.SetProperty() formulae still work. They seem to have changed to GetPropertyCaller.Components.GetProperty?by_Object_String() in this PR.

**Comment by rwemay** [comment]
@kayleighhoude , good if you can help test this to get it over the line - we need to make sure all of the sheets you're using currently work with this PR, as well as the issues identified.

**Review by kayleighhoude — COMMENTED** [review]
Ran tests on 3 separate workflows:
- [x] Enum menu appears
- [x] Internalise retaining information between sessions (woohoo!)
- [x] Gets and sets retaining their information
- [x] Reopen on crash not happening anymore

1 workbook has 4 sheets with pulls on each for about 50 objects, and upon opening and attempting to set a pull to false/true the program fails (this is a consistent issue)
image
^or upon opening this runs over and over again

**Comment by rwemay** [comment]
Does Engine.GetProperty still work for you on this PR @kayleighhoude ?

**Comment by kayleighhoude** [comment]
@rwemay it was working on all of the Revit properties that I was getting in those workflows

**Comment by kayleighhoude** [comment]
@awakeman heads up that i'm working off of the same branch now (post cleaning up my build folder and moving fresh Excel UI, Engine and oM dlls over to BHoM in the Roaming folder)

**Comment by rwemay** [comment]
For the final updates to this PR, can we also set copy local on the BHoM_Engine to false? 

https://github.com/BHoM/Excel_Toolkit/issues/227

**Comment by rwemay** [comment]
@al-fisher , @IsakNaslundBh , I've pushed up the changes discussed, centralising a lot of the formula name code to the Caller template. This is currently dependent on the branch in BHoM_UI just below - but as we said, hoping to get that merged ahead so no need to rename branches/rePR etc. 

https://github.com/BHoM/BHoM_UI/tree/BHoM_UI-%23272-Add-GetFullName()-method-for-callers

**Comment by rwemay** [comment]
@awakeman @IsakNaslundBh @kayleighhoude @al-fisher need to look at the adapter formulae as with my last push there's a bit of repetition. 

@kayleighhoude @awakeman can you build my latest and test - see what you think of the formulae - essentially adding BH.Engine or BH.oM across the board to keep things 1:1. The post builds should work now such that you _don't_ need to build BHoM_UI - only this toolkit/branch for testing (thanks @IsakNaslundBh ). We need to get this fully tested and merged by Tuesday next week(!!)

File: Excel_UI/Excel_UI.csproj [hunk 1]
```diff
@@ -40,6 +40,10 @@
       <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM_Adapter.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="BHoM_Engine">
+      <HintPath>..\..\..\..\..\ProgramData\BHoM\Assemblies\BHoM_Engine.dll</HintPath>
```
Comment (IsakNaslundBh) [inline Excel_UI/Excel_UI.csproj]: this needs to be changed to exact path

**Review by IsakNaslundBh — DISMISSED** [review]
wrong path for BHoM_Engine

File: Excel_UI/Excel_UI.csproj (hunk 1 above)
Comment (awakeman) [reply Excel_UI/Excel_UI.csproj]: Thanks, I spotted just after too.

**Review by rwemay — APPROVED** [review]
Approving on the basis of having (hopefully) fixed the crash on close but. Concious that there are issues on formulae which we will raise and squash today with any luck. If not then apologies for breaking changes on formulae in the alpha installers/master branch...

**Pull request merged**